sdk/src/math/amm.ts

```diff
@@ -25,25 +27,86 @@ import {
 	calculateAdjustKCost,
 	calculateBudgetedPeg,
 } from './repeg';
-export function calculateNewAmm(
+
+export function calculateOptimalPegFromTargetPrice(
+	targetPrice: BN,
+	baseAssetReserve: BN,
+	quoteAssetReserve: BN
+): BN {
+	return targetPrice
+		.mul(baseAssetReserve)
+		.div(quoteAssetReserve)
+		.add(MARK_PRICE_PRECISION.div(PEG_PRECISION).div(new BN(2)))
+		.div(MARK_PRICE_PRECISION.div(PEG_PRECISION));
+}
+
+export function calculateOptimalPegAndBudget(
 	amm: AMM,
 	oraclePriceData: OraclePriceData
-): [BN, BN, BN, BN] {
-	let pKNumer = new BN(1);
-	let pKDenom = new BN(1);
-
+): [BN, BN, BN, boolean] {
+	const markPriceBefore = calculatePrice(
+		amm.baseAssetReserve,
+		amm.quoteAssetReserve,
+		amm.pegMultiplier
+	);
 	const targetPrice = oraclePriceData.price;
-	let newPeg = targetPrice
+	const newPeg = targetPrice
 		.mul(amm.baseAssetReserve)
 		.div(amm.quoteAssetReserve)
 		.add(MARK_PRICE_PRECISION.div(PEG_PRECISION).div(new BN(2)))
 		.div(MARK_PRICE_PRECISION.div(PEG_PRECISION));
-	let prePegCost = calculateRepegCost(amm, newPeg);
+	const prePegCost = calculateRepegCost(amm, newPeg);
 
 	const totalFeeLB = amm.totalExchangeFee.div(new BN(2));
 	const budget = BN.max(ZERO, amm.totalFeeMinusDistributions.sub(totalFeeLB));
+	if (budget.lt(prePegCost)) {
+		const maxPriceSpread = new BN(amm.maxSpread)
+			.mul(targetPrice)
+			.div(BID_ASK_SPREAD_PRECISION);
+
+		// console.log('maxPriceSpread:', maxPriceSpread.toNumber());
+
+		let markAdj: BN;
```

Comment:
i dont think you need to define these all here since you just return the if block below